Allow unconstrained instances & disallow overlapping instances

jespercockx opened the pull request:
This PR concerns two related but separate changes to instance argument resolution.

**1. Allow unconstrained instances** 
The current restriction on instance arguments requires there to be no non-rigid metavariables in the type. However, this turns out to be too restrictive in practice (see #723 and #2760). I discussed with @UlfNorell but there seems to be no reasonable way to loosen the restriction and still en up with something that's understandable to the user. So I propose to drop the restriction altogether. This has two adverse consequences:

- Possibly looping / exponential behaviour in the presence of instance arguments of unknown/partially known type.
- Possibly unpredictable behaviour where there's only a single instance in scope 'by accident', e.g. a constraint `{{_ : Eq _1}}` gets solved with `instance EqBool : Eq Bool`, setting `_1 = Bool` in the process.

The second change in this PR is an attempt to significantly reduce these two negative effects (however, it doesn't eliminate them completely).

**2. Disallow overlapping instances** 
The problem with the looping/exponential behaviour is caused by the fact that Agda allows *overlapping instances*. Overlapping instances are known by the Haskell community to be a Bad Thing (tm). Technically, two instances overlap if they could both solve the instance goal *when given appropriate solutions for their recursive (instance) arguments*. So the only thing that makes Agda choose one over the other is that for one of them, recursive instance search produces no solutions. But this failing search could go arbitrarily deep, as well as branch arbitrarily many times, thus causing the looping/exponential behaviour. 

The second change in this PR is thus to disallow overlapping instances by default by postponing recursive instance search until after Agda has committed to a specific instance. (They can be enabled again with the  `--overlapping-instances` flag). This has a few good consequences:

- It makes recursive instance search *linear* in the depth of the instance (up to `--max-instance-depth` in the worst case). It should hence also fix #2993. (EDIT: it does!)
- It makes it easier to reason *locally* about when Agda will pick a certain instance.
- Adding one instance will no longer make another instance 'higher up' in the search tree fail, so instance search becomes somewhat more robust under the addition of new instances.

I tested this proposed change on the test suite (some breakage but overall not too bad, see code for details) and on Ulf's prelude (no breakage).

jespercockx committed:
[ re #723 ] Remove outdated comment

jespercockx committed:
[ re #723 ] Remove restriction on unconstrained metas for instance search

jespercockx committed:
[ instance ] Disable overlapping instances by default

Technically, we postpone recursive instance search until after we have committed to a solution

Overlapping instances can optionally be enabled with a flag.

jespercockx committed:
[ fix #2993 ] Added test case

jespercockx committed:
[ fix #2670 ] Added test case

jespercockx commented:
Since when does Github show all the force-pushes in a PR? Now all the dirty secrets of my workflow are exposed :(

nad commented:
I subjected your code to one of my repositories, and for the first file with problems the error buffer contains about 2400 lines...

If you want to try it yourself:
```
git clone http://www.cse.chalmers.se/~nad/repos/equality/
(cd equality && git checkout c8b6831ab6a447780b21fc8c0e1fad7c181cbb98)
git clone http://www.cse.chalmers.se/~nad/repos/delay-monad/
(cd delay-monad && git checkout 04df7090e6b69a9e0d6733fb30e284f7a31785aa)
git clone http://www.cse.chalmers.se/~nad/repos/up-to/
(cd up-to && git checkout 112fb59148aa79c06276534901663ae90c30e131)
echo $PWD/equality/equality.agda-lib > libraries
echo $PWD/delay-monad/delay-monad.agda-lib >> libraries
echo $PWD/up-to/up-to.agda-lib >> libraries
cd up-to
agda -i. --no-default-libraries --library-file=../libraries README.agda
```

jespercockx committed:
[ instances ] Don't turn explicit arguments into instance arguments during instance search

jespercockx commented:
Right, there's still a problem that if you have a parametrized module/record with an instance in it and later apply the module, there are two copies of the instance that are considered overlapping. This was actually already a problem before this patch, but now it shows up much more frequently because we can no longer rule out the parametrized instances.

I think the best thing to do here is to no longer automatically turn explicit arguments into instance arguments for the purpose of instance search. So an instance declared in a parametrized module would no longer use instance search to resolve the (explicit) parameters. 

I tried this out on your library and it seems to fix the problem you encountered. It should hopefully also fix other problems with instances in parametrized modules/records. I just need to add a warning for when the user accidentally declares an instance with explicit arguments, and then I'll update the PR.

jespercockx committed:
[ instance ] If instance meta is irrelevant, also allow overlapping 'maybe' solutions

jespercockx committed:
[ instance ] Warn when declaring an instance with explicit arguments

jespercockx committed:
[ re #1003 ] Don't ignore instance constraint on already solved instance meta

jespercockx committed:
[ instance ] Document changes to instance resolution in the changelog

jespercockx committed:
[ instance ] Document overlapping instances in the user manual

jespercockx commented:
I'm trying these changes on the `HoTT-Agda` library but it seems to require quite a few (small-ish) changes, because there are many overlapping instances, as well as many instances with explicit arguments. I made some progress but after a while I got stuck on `Truncation.agda` (see https://github.com/jespercockx/HoTT-Agda/commit/a15f184df73304a7bfeeaac96194b89286501bdb for my WIP). 

Perhaps @favonia or @guillaumebrunerie would want to take a look and give their opinion on these changes to instance resolution?

jespercockx committed:
[ fix #3175 ] Added test case

jespercockx commented:
I'll merge this now, but please let me know if you have any more suggestions for improvement!

m0davis commented:
The code block in `Overlapping Instances` is not displaying (at least on github). Maybe you need to add a colon (i.e. `.. code-block:: agda`)?

nad commented on doc/user-manual/language/instance-arguments.lagda.rst:
```diff
@@ -439,6 +439,36 @@ must be implicit, indicating that it needs to be inferred by
+are defined to overlap if they could both solve the instance goal
+when given appropriate solutions for their recursive (instance)
+arguments.
+
```
It might be worth pointing out that, even if overlapping instances are allowed, Agda will not make unforced choices.

identicalsnowflake commented:
@jespercockx 

Mm, this change really kills [some of the stuff I was doing previously with instance search](https://identicalsnowflake.github.io/InstanceSearch.html) (it's basically just saying given any big stack of sums/products, if there's a unique way to project out a smaller stack of sums and products, I'll find it for you).

I can turn on overlapping instances, but previously, that stuff type checked fairly quickly, but now, I can't get even simple examples to ever finish type checking, even with fairly low max instance search depth.

Is there anything I could change or any option I could set to get this to be useable again? Because this was... quite useful.

jespercockx commented:
I'm not sure what to say... The old code (i.e. the restriction on unconstrained instances) was causing a lot of problems and was actively preventing me from using instance arguments. @UlfNorell and I discussed how to limit the restriction to fix these issues but we didn't really find any sensible solution, so I removed it entirely and added the new restriction on overlapping instances instead to prevent looping behaviour.

I guess we could bring back the old code under a flag (and perhaps enable this flag automatically when `--overlapping-instances` is enabled) but then we need someone who wants to maintain this code -- I have other things I'd rather work on.

identicalsnowflake commented:
Ah, well if it's going to be a pain, don't worry about it. I'll figure something else out.
